tsdb/head_append.go

```diff
@@ -1391,32 +1414,93 @@ func (a *headAppender) Commit() (err error) {
 			a.head.metrics.chunksCreated.Inc()
 		}
 
-		series.cleanupAppendIDsBelow(a.cleanupAppendIDsBelow)
-		series.pendingCommit = false
-		series.Unlock()
+		acc.series.cleanupAppendIDsBelow(a.cleanupAppendIDsBelow)
+		acc.series.pendingCommit = false
+		acc.series.Unlock()
 	}
+}
 
+// commitMetadata commits the metadata for each series in the headAppender.
+// It iterates over the metadata slice and updates the corresponding series
+// with the new metadata information. The series is locked during the update
+// to ensure thread safety.
+func (a *headAppender) commitMetadata(acc *appenderCommitContext) {
 	for i, m := range a.metadata {
-		series = a.metadataSeries[i]
-		series.Lock()
-		series.meta = &metadata.Metadata{Type: record.ToMetricType(m.Type), Unit: m.Unit, Help: m.Help}
-		series.Unlock()
+		acc.series = a.metadataSeries[i]
+		acc.series.Lock()
+		acc.series.meta = &metadata.Metadata{Type: record.ToMetricType(m.Type), Unit: m.Unit, Help: m.Help}
+		acc.series.Unlock()
+	}
+}
+
+// Commit writes to the WAL and adds the data to the Head.
+// TODO(codesome): Refactor this method to reduce indentation and make it more readable.
+func (a *headAppender) Commit() (err error) {
+	if a.closed {
+		return ErrAppenderClosed
+	}
+	defer func() { a.closed = true }()
+
+	if err := a.log(); err != nil {
+		_ = a.Rollback() // Most likely the same error will happen again.
+		return fmt.Errorf("write to WAL: %w", err)
+	}
+
+	if a.head.writeNotified != nil {
+		a.head.writeNotified.Notify()
+	}
+
+	a.commitExemplars()
+
+	defer a.head.metrics.activeAppenders.Dec()
+	defer a.head.putAppendBuffer(a.samples)
+	defer a.head.putSeriesBuffer(a.sampleSeries)
+	defer a.head.putExemplarBuffer(a.exemplars)
+	defer a.head.putHistogramBuffer(a.histograms)
+	defer a.head.putFloatHistogramBuffer(a.floatHistograms)
+	defer a.head.putMetadataBuffer(a.metadata)
+	defer a.head.iso.closeAppend(a.appendID)
+
+	acc := &appenderCommitContext{
+		floatsAppended:     len(a.samples),
+		histogramsAppended: len(a.histograms) + len(a.floatHistograms),
+		inOrderMint:        math.MaxInt64,
+		inOrderMaxt:        math.MinInt64,
+		oooMinT:            math.MaxInt64,
+		oooMaxT:            math.MinInt64,
+		oooCapMax:          a.head.opts.OutOfOrderCapMax.Load(),
+		appendChunkOpts: chunkOpts{
+			chunkDiskMapper: a.head.chunkDiskMapper,
+			chunkRange:      a.head.chunkRange.Load(),
+			samplesPerChunk: a.head.opts.SamplesPerChunk,
+		},
 	}
```

Comment:
Nice, I like the idea of creating an object for this. Maybe we could have a `NewCommitContext` function to decrease lines a little bit more?